lte/gateway/c/li_agent/PDUGenerator.cpp

```diff
@@ -0,0 +1,155 @@
+    return true;
+  }
+  return false;
+}
+
+}  // namespace
+
+namespace magma {
+
+PDUGenerator::PDUGenerator(
+    std::unique_ptr<ProxyConnector> proxy_connector,
+    std::unique_ptr<DirectorydClient> directoryd_client,
+    const std::string& pkt_dst_mac, const std::string& pkt_src_mac)
+    : pkt_dst_mac_(pkt_dst_mac),
+      pkt_src_mac_(pkt_src_mac),
+      directoryd_client_(std::move(directoryd_client)),
+      proxy_connector_(std::move(proxy_connector)) {}
+
+void PDUGenerator::set_conditional_attr(
+    const struct pcap_pkthdr* phdr, struct conditional_attributes* attributes) {
+  attributes->timestamp = phdr->ts.tv_sec;
+}
+
+void* PDUGenerator::generate_pkt(
+    const struct pcap_pkthdr* phdr, const u_char* pdata) {
+  uint8_t* data        = static_cast<uint8_t*>(calloc(
+      1, sizeof(struct pdu_info) + sizeof(struct conditional_attributes) +
+             phdr->len));
+  struct pdu_info* pdu = (struct pdu_info*) data;
+
+  pdu->version  = PDU_VERSION;
+  pdu->pdu_type = PDU_TYPE;
+  pdu->header_length =
+      sizeof(struct pdu_info) + sizeof(struct conditional_attributes);
+  pdu->payload_length = phdr->len;
+  pdu->payload_format = IP_PAYLOAD_FORMAT;
+
+  set_conditional_attr(
+      phdr, (struct conditional_attributes*) (data + sizeof(struct pdu_info)));
+
+  memcpy(data + pdu->header_length, pdata, phdr->len);
+
+  MLOG(MDEBUG) << "Generated packet length with length - "
+               << pdu->header_length + pdu->payload_length;
+  print_bytes(data, pdu->header_length + pdu->payload_length);
+
+  return (void*) data;
+}
+
+bool PDUGenerator::send_packet(
+    const struct pcap_pkthdr* phdr, const u_char* pdata) {
+  std::string src_ip;
+  std::string dst_ip;
+  if (!extract_ip_addr(pdata, src_ip, dst_ip)) {
+    MLOG(MERROR) << "Could not extract IP from the packet, skipping";
+    return true;
+  }
+
+  MLOG(MDEBUG) << "Processing packet with src ip " << src_ip << " dst ip "
+               << dst_ip;
+  void* data           = generate_pkt(phdr, pdata);
+  struct pdu_info* pdu = (struct pdu_info*) data;
+
+  directoryd_client_->get_directoryd_xid_field(
+      dst_ip, std::bind(
+                  &PDUGenerator::handle_ip_lookup_callback, this, src_ip, data,
+                  pdu, std::placeholders::_1, std::placeholders::_2));
+
+  directoryd_client_->get_directoryd_xid_field(
+      dst_ip, std::bind(
+                  &PDUGenerator::handle_ip_lookup_callback, this, dst_ip, data,
+                  pdu, std::placeholders::_1, std::placeholders::_2));
+  return true;
+}
+
+void PDUGenerator::handle_ip_lookup_callback(
+    std::string ip_addr, void* data, struct pdu_info* pdu, Status status,
+    DirectoryField resp) {
+  // TODO process the output of the directoryd lookup
+  pdu->payload_direction = DIRECTION_TO_TARGET;
+  set_xid(pdu, "tracking_123");
+  proxy_connector_->send_data(data, pdu->header_length + pdu->payload_length);
+  // free(data);
```

Comment:
Why is this commented out? Should we free `data` here ?

Reply:
yup, just didn't commit yet